contracts/debtAllocators/YieldManager/YieldManager.sol

```diff
@@ -0,0 +1,388 @@
+        uint256 _accountedFor = IVault(_vault).totalIdle();
+        // Create local variables used through loops.
+        address _strategy;
+        uint256 _currentDebt;
+        uint256 _newDebt;
+        for (uint256 i = 0; i < _newAllocations.length; ++i) {
+            _strategy = _newAllocations[i].strategy;
+            _newDebt = uint256(_newAllocations[i].newDebt);
+            // Get the debt the strategy current has.
+            _currentDebt = IVault(_vault).strategies(_strategy).current_debt;
+            // Add to what we have accounted for.
+            _accountedFor += _currentDebt;
+
+            // Get the current weighted APR the strategy is earning
+            uint256 _strategyApr = (aprOracle.getStrategyApr(_strategy, 0) *
+                _currentDebt);
+
+            // Add to the amount currently being earned.
+            _currentYield += _strategyApr;
+
+            // If we are withdrawing.
+            if (_currentDebt > _newDebt) {
+                // If we are pulling all debt from a strategy.
+                if (_newDebt == 0) {
+                    // We need to report profits and have them immediately unlock to not lose out on locked profit.
+                    Keeper(keeper).report(_strategy);
+                }
+
+                if (
+                    // We cannot decrease debt if the strategy has any unrealised losses.
+                    IVault(_vault).assess_share_of_unrealised_losses(
+                        _strategy,
+                        _currentDebt
+                    ) != 0
+                ) {
+                    // Realize the loss.
+                    IVault(_vault).process_report(_strategy);
+                }
+            }
+
+            // Get the target based on the new debt.
+            uint256 _targetRatio = _newDebt < _totalAssets
+                ? (_newDebt * MAX_BPS) / _totalAssets
+                : MAX_BPS;
+
+            // If different than the current target.
+            if (
+                GenericDebtAllocator(allocator).getStrategyTargetRatio(
+                    _strategy
+                ) != _targetRatio
+            ) {
+                // Update allocation.
+                GenericDebtAllocator(allocator).setStrategyDebtRatios(
+                    _strategy,
+                    _targetRatio
+                );
+            }
+
+            // Add the expected change in APR based on new debt.
+            if (_newDebt == _currentDebt) {
+                // We assume the new apr will be the same as current.
+                _afterYield += _strategyApr;
+            } else if (_newDebt != 0) {
+                _afterYield += (aprOracle.getStrategyApr(
+                    _strategy,
+                    int256(_newDebt) - int256(_currentDebt)
+                ) * _newDebt);
+            }
+        }
+
+        // Make sure the ending amounts are the same otherwise rates could be wrong.
+        require(_totalAssets == _accountedFor, "cheater");
```

Comment:
Might be worth adding some dev messages `// dev: wtf`

Reply:
i think "cheater" is unique enogh to find out where its reverting